File: packages/shared/src/schemas/feature-flags.ts
import { z } from "zod";

export const FeatureFlagKeySchema = z.enum([
  "invitations",
  "business_claiming",
  "weekly_digest",
  "community_posting",
  "business_reviews",
  "user_registration",
]);
export const FeatureFlagScopeSchema = z.enum(["GLOBAL", "USER", "COMPOUND", "CITY", "NEIGHBOURHOOD"]);

export const FeatureFlagSchema = z.object({
  key: FeatureFlagKeySchema,
  enabled: z.boolean(),
  description: z.string().nullable().optional(),
  updated_at: z.string().datetime().optional(),
});

export const FeatureFlagOverrideSchema = z.object({
  id: z.number().int().positive().optional(),
  key: FeatureFlagKeySchema,
  scope: FeatureFlagScopeSchema,
  scope_id: z.string().min(1).nullable().optional(),
  enabled: z.boolean(),
  starts_at: z.string().datetime().nullable().optional(),
  ends_at: z.string().datetime().nullable().optional(),
});

export const FeatureConfigSchema = z.object({
  flags: z.record(FeatureFlagKeySchema, z.boolean()),
  city_enabled: z.boolean().optional(),
  neighbourhood_enabled: z.boolean().optional(),
  fetched_at: z.string().datetime().optional(),
});

export const FeatureFlagUpdateSchema = z.object({
  enabled: z.boolean(),
  description: z.string().trim().max(500).optional(),
});

export type FeatureFlagKey = z.infer<typeof FeatureFlagKeySchema>;
export type FeatureFlagScope = z.infer<typeof FeatureFlagScopeSchema>;
export type FeatureFlag = z.infer<typeof FeatureFlagSchema>;
export type FeatureFlagOverride = z.infer<typeof FeatureFlagOverrideSchema>;
export type FeatureConfig = z.infer<typeof FeatureConfigSchema>;
export type FeatureFlagUpdate = z.infer<typeof FeatureFlagUpdateSchema>;

File: backend/app/schemas/feature_flags.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


FeatureFlagKey = Literal[
    "invitations",
    "business_claiming",
    "weekly_digest",
    "community_posting",
    "business_reviews",
    "user_registration",
]
FeatureFlagScopeValue = Literal["USER", "COMPOUND", "CITY"]


class FeatureFlagWrite(BaseModel):
    enabled: bool
    description: str | None = Field(default=None, max_length=500)
    config: dict[str, Any] = Field(default_factory=dict)


class FeatureFlagCreate(FeatureFlagWrite):
    key: FeatureFlagKey


class FeatureFlagResponse(FeatureFlagCreate):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime


class FeatureFlagOverrideWrite(BaseModel):
    scope: FeatureFlagScopeValue
    enabled: bool
    user_id: int | None = Field(default=None, gt=0)
    compound_id: int | None = Field(default=None, gt=0)
    city: str | None = Field(default=None, min_length=1, max_length=120)
    config: dict[str, Any] = Field(default_factory=dict)

    @model_validator(mode="after")
    def validate_scope_target(self):
        expected = {
            "USER": self.user_id,
            "COMPOUND": self.compound_id,
            "CITY": self.city,
        }
        if expected[self.scope] is None or sum(value is not None for value in expected.values()) != 1:
            raise ValueError("Exactly the target matching scope must be supplied")
        return self


class FeatureFlagOverrideResponse(FeatureFlagOverrideWrite):
    model_config = ConfigDict(from_attributes=True)

    id: int
    feature_flag_id: int
    target_key: str
    created_at: datetime
    updated_at: datetime


class FeatureConfigResponse(BaseModel):
    flags: dict[FeatureFlagKey, bool]
    city_enabled: bool
    neighbourhood_enabled: bool
    fetched_at: datetime
